### PR TITLE
NAS-102186 / 12.0 / Nas 102186

### PR DESCRIPTION
**Pull request by dkmullen**
Adds two-factor auth component under the System heading
- Right now, 2FA for SSH from the command line doesn't work (Waqar is investigating) so don't set it
- Safety tip: Make sure your system clock is correctly set
- You have to enable 2FA to get a 'secret'. This might need to be made clearer through labeling???
- If you're secret is null or '', you shouldn't be able to view the QR code.
- You can add your system to your authenticator app via the QR code or by manually entering the secret.
PS  There is a new module to be added too.

### CODE REVIEW

File: src/app/pages/system/system.routing.ts [hunk 1]
```diff
@@ -344,6 +346,11 @@ export const routes: Routes = [
           data: { title: 'EULA', breadcrumb: 'EULA'}
         }
       ]
+    },
+    {
+      path: 'two-factor',
+      component: TwoFactorComponent,
+      data: { title: 'Two-Factor Auth', breadcrumb: 'Two-Factor Auth' },
```
Comment (erinix) [inline src/app/pages/system/system.routing.ts]: the title and breadcrumb should technically have T() on them too

**Comment by erinix** [comment]
please review the tooltips @freenas/docs-team

**Comment by erinix** [comment]
You'll need to fix your merge conflict :)

File: src/app/helptext/system/2FA.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    placeholder: T('One-Time Password (OTP) Digits'),
+    tooltip: T('The number of digits in the One-Time Password. The default value is 6, \
+ which is the length of the standard OTP from Google. Check the settings of your app or device \
+ before selecting this.'),
+    validation: [Validators.min(6), Validators.max(8)]
+  },
+  interval: {
+    placeholder: T('Interval'),
+    tooltip: T('The lifespan (in seconds) of each One-Time Password. Default is 30 seconds. \
+ The minimum value is 5.'),
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
 The minimum lifetime is 5 seconds.'),
```

File: src/app/helptext/system/2FA.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    validation: [Validators.min(5)]
+  },
+  window: {
+    placeholder: T('Window'),
+    tooltip: T('Use <i>window</i> to extend the validity of passwords beyond the <i>interval</i> setting. \
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
    tooltip: T('Use <i>Window</i> to extend the validity of passwords beyond the <i>Interval</i> setting. \
```

File: src/app/helptext/system/2FA.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+A window setting of 1, for example, means that one password before and after the current one is valid, \
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
For example, a window setting of 1 means that one password before and after the current one is valid. \
```

File: src/app/helptext/system/2FA.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+in addition to the current password. Extending the window can be useful in high-latency situations. \
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
leaving three valid passwords. Extending the window can be useful in high-latency situations. \
```

File: src/app/helptext/system/2FA.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+IMPORTANT: Two-factor authentication is time-based, and requires that the system\'s time is set correctly.'),
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
IMPORTANT: Two-factor authentication is time-based and requires that the system time is set correctly.'),
```

File: src/app/helptext/system/2FA.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  services: {
+    placeholder: T('Enable Two-Factor Auth for SSH'),
+    tooltip: T('Enable two-factor authentication for SSH access to the system. It is recommended \
+ to leave this DISABLED until after two-factor authentication is successfully tested with the UI.')
+  }, 
+
+  secret: {
+    placeholder: T('Secret (Read only)'),
+    tooltip: T('The secret used to generate OTPs. The secret is produced by the system when Two-Factor \
+ Authentication is first activated, so this field is read-only.')
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
 Authentication is first activated.')
```

File: src/app/helptext/system/2FA.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  },
+ 
+  uri: {
+    placeholder: T('Provisioning URI (includes Secret - Read only):'),
+    tooltip: T('The URI used to provision an OTP. The URI (which contains the secret) is encoded in a QR Code. \
+ To set up an OTP app like Google Authenticator, use the app to scan the QR code or enter the secret manually \
+ into the app. The URI is produced by the system when Two-Factor Authentication is first activated, \
+ so this field is read-only.')
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
 
```

File: src/app/helptext/system/2FA.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    placeholder: T('Secret (Read only)'),
+    tooltip: T('The secret used to generate OTPs. The secret is produced by the system when Two-Factor \
+ Authentication is first activated, so this field is read-only.')
+  },
+ 
+  uri: {
+    placeholder: T('Provisioning URI (includes Secret - Read only):'),
+    tooltip: T('The URI used to provision an OTP. The URI (which contains the secret) is encoded in a QR Code. \
+ To set up an OTP app like Google Authenticator, use the app to scan the QR code or enter the secret manually \
+ into the app. The URI is produced by the system when Two-Factor Authentication is first activated, \
```
Comment (Mrt134) [inline src/app/helptext/system/2FA.ts]:
```suggestion
 into the app. The URI is produced by the system when Two-Factor Authentication is first activated.')
```

**Pull request merged**